rosette/src/Import.cc

```diff
@@ -440,11 +462,111 @@ pOb createRosetteObject(ObjectCodePB::Object * ob) {
+    }
+    
+    case ObjectCodePB::OT_ATOMIC_DESCRIPTOR: {
+        ObjectCodePB::AtomicDescriptor * pbob = ob->mutable_atomic_descriptor();
+
+        AtomicDescriptor * ad = AtomicDescriptor::create();
+        ad->_offset = pbob->_offset();
+        ad->_align_to = pbob->_align_to();
+        ad->_size = pbob->_size();
+        ad->mnemonic = createRosetteObject(pbob->mutable_mnemonic());
+        ad->imported = createRosetteObject(pbob->mutable_imported());
+        ad->freeStructOnGC = createRosetteObject(pbob->mutable_freestructongc());
+        ad->_signed = (RblBool *)createRosetteObject(pbob->mutable__signed());
+
+        retval = ad;
+        break;
+    }
+    
+    case ObjectCodePB::OT_REFLECTIVE_MTHD: {
+        ObjectCodePB::ReflectiveMthd * pbob = ob->mutable_reflective_mthd();
+        retval = ReflectiveMthd::create( (Code *)createRosetteObject(pbob->mutable_code()),
+                                                createRosetteObject(pbob->mutable_id()),
+                                                createRosetteObject(pbob->mutable_source()));
+
+        break;
+    }
+    
+    case ObjectCodePB::OT_PRODUCT_TYPE: {
+        ObjectCodePB::ProductType * pbob = ob->mutable_product_type();
+
+        Tuple * tup = Tuple::create(pbob->elements_size(), INVALID);
+        for (int i = 0; i < pbob->elements_size(); i++) {
+            tup->setNth( i, createRosetteObject(pbob->mutable_elements(i)) );
+        }
+        retval = ProductType::create( tup, NIV );
+        break;
+    }
+    
+    case ObjectCodePB::OT_SUM_TYPE: {
+        ObjectCodePB::SumType * pbob = ob->mutable_sum_type();
+
+        Tuple * tup = Tuple::create(pbob->elements_size(), INVALID);
+        for (int i = 0; i < pbob->elements_size(); i++) {
+            tup->setNth( i, createRosetteObject(pbob->mutable_elements(i)) );
+        }
+        retval = SumType::create( tup );
+        break;
+    }
+    
+    case ObjectCodePB::OT_MULTI_METHOD: {
+        ObjectCodePB::MultiMethod * pbob = ob->mutable_multi_method();
+
+        MultiMethod * mm = MultiMethod::create();
+
+        // A StdExtension is similar to a tuple which is a list of objects.
+        Tuple * tup = Tuple::create(pbob->elements_size(), INVALID);
+        for (int i = 0; i < pbob->elements_size(); i++) {
+            BASE(tup)->setNth( i, createRosetteObject(pbob->mutable_elements(i)) );
+        }
+        mm->extension = StdExtension::create( tup );
+
+        retval = mm;
+        break;
+    }
+
+    case ObjectCodePB::OT_STD_OPRN: {
+        ObjectCodePB::StdOprn * pbob = ob->mutable_std_oprn();
+
+        // TODO: fix these parameters
+        retval = StdOprn::create(INVALID, RBLBOOL(false));
+    }
+
     default:
         warning("Import object type=%d not yet implemented!\nObjectPB=\n%s\n", ob->type(), ob->DebugString().c_str());
         assert(false);
     }
 
+    // Handle meta and parent fields. Not for RblAtom objects.
+    if (retval && TAG(retval) == OTptr) {
```

Comment:
`RblAtom`s have `meta` and `parent` objects as well. Is this handled elsewhere?

Reply:
The only place that RblAtom objects are fully instantiated that I know of is in RblAtom.cc around line 323. Here a single instance of each is created for use in the BASE() function which calls decodeAtom(). This function copies the meta data into the single instance of the object, and returns a pointer to it.  This seems to be only used to call member functions, that print, display, or otherwise output the value of the Atom.  The rest of the time, as far as I know, RblAtoms are passed around as pseudo objects with their data contained within the pOb itself. It this use, there are no meta or parent objects.

Reply:
`Fixnum` is a `RblAtom` and has a `meta` and `parent` for example.

Reply:
Yes, but once you called `BASE()` on the pointer and for example got a `Fixnum`, that `Fixnum` has a `meta` and a `parent` we have to keep track of when exporting.

Reply:
The only meta and parent that it has at that point are those of the global "prototypicalFixnum" object. As near as I can tell, those are never set to anything interesting.

There is really only ever one instance of the Fixnum object, and that is the "prototypicalFixnum" created by MODULE_INIT() in RblAtom.cc. As near as I can tell, the meta and parent pointers there are never set to anything different for a specific Fixnum.

Look closely at how decodeAtom() in RblAtom.cc works. You'll see that the only thing that happens when it is called is the pseudo-pointer pOb tag is referenced to determine what type of RblAtom this is, and then the pseudo-pointer is transferred into the atom member variable of the appropriate global Atom instance, and the address of the global is returned.

The bottom line is, every time you use BASE() to access a Fixnum, you will get the same memory address pointing to the same Fixnum instance, with the same meta and parent pointers. The only thing that changes is the value of the "atom" member variable will contain the original pOb contents that were passed to BASE().

Reply:
They are of interested. For example the `Fixnum` parent is the `Fixnum SBO` which maps from operations (e.g. the `+` operation/`StdOprn`) to primitives (e.g. the `fx+` primitive).
These mappings get built up while running `boot.rbl`.

I guess you are right in the sense that it's not really important for exporting and importing. All we have to know is the type and the rest can be hardcoded.

Reply:
Yes. What you described there sounds fine.